### PR TITLE
[Feature] Counties Coverage Chart

### PR DESCRIPTION
**Pull request by gertie-sheshe**
## Description

Use live data to populate the County Coverage chart.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Screenshots

Screenshot from 2021-06-22 16-24-22

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: public/js/counties.js [hunk 1]
```diff
@@ -471,15 +471,32 @@ const countiesLocation = {
...
+      $('.menu-button').click();
+    }
+  });
+  Object.keys(countiesLocation).forEach((key) => {
+    const c = new Option(countiesLocation[key].label, key);
+    countySelect.options.add(c);
+  });
+
+  // fetch data. Blocking JS? + quite slow
+  const data = await fetch('/api/data');
```
Comment (kilemensi) [inline public/js/counties.js]: Include from when should data be fetched here i.e. compute the date of 7 days ago from today here.

File: public/js/counties.js [hunk 2]
```diff
@@ -471,15 +471,32 @@ const countiesLocation = {
...
+  const results = await data.json();
+
+  // filter Nairobi. This will be used to populate the chart when counties are available
+  const nairobiCounty = results.results.filter(
+    (result) => result.location.city === 'Nairobi'
+  );
+
+  // no of nodes in Nairobi and total nodes
+  const nairobiNodes = nairobiCounty.length;
+  const totalNodes = results.count;
```
Comment (kilemensi) [inline public/js/counties.js]: All these can be done in the changeCountyGraph method, we just pass in data and the name of county.

File: public/js/events.js [hunk 3]
```diff
@@ -3,7 +3,8 @@ const handleLocationChange = async (value) => {
   const newUrl = `${baseUrl}${countiesLocation[value].zoom}/${countiesLocation[value].latitude}/${countiesLocation[value].longitude}`;
 
   const county = countiesLocation[value].name;
-  const numberOfHouseholds =await getHouseHoldCount(county);
+  countyGraphChange(county);
```
Comment (kilemensi) [inline public/js/events.js]: Where would this get data from?

File: src/api/index.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import fetch from 'isomorphic-unfetch';
 import Papa from 'papaparse';
 import request from 'request';
+// import nodesData from './data';
```
Comment (kilemensi) [inline src/api/index.js]: ?

**Review by kilemensi — COMMENTED** [review]
👍🏾 


-----


Why does the chart show **Others**? Don't have list of all counties / cities?

**Review by nerdlet — APPROVED** [review]
Look good to me!

File: public/js/counties.js (hunk 2 above)
Comment (gertie-sheshe) [reply public/js/counties.js]: Fixed

File: src/api/index.js (hunk 4 above)
Comment (gertie-sheshe) [reply src/api/index.js]: This was the data I used temporarily. Removed it

File: public/js/events.js (hunk 3 above)
Comment (gertie-sheshe) [reply public/js/events.js]: Refactored the `countyGraphChange` method to have access to the data @kilemensi

File: public/js/counties.js (hunk 1 above)
Comment (gertie-sheshe) [reply public/js/counties.js]: Done

File: public/js/counties.js [hunk 5]
```diff
@@ -470,20 +470,29 @@ const countiesLocation = {
...
+    const c = new Option(countiesLocation[key].label, key);
+    // TODO(Brenda): We only have Nairobi data so we'll disable all other counties for now
+    if (c.value !== 'nairobi') {
+      c.disabled = true;
+    }
+    countySelect.options.add(c);
+  });
+
+  // fetch data. Blocking JS? + quite slow
+  const data = await fetch('/api/data/?length=7');
```
Comment (kilemensi) [inline public/js/counties.js]:
```suggestion
  const data = await fetch('/api/data/?days=7');
```

File: public/js/counties.js [hunk 6]
```diff
@@ -470,20 +470,29 @@ const countiesLocation = {
...
+  results = await data.json();
+
+  countyGraphChange('Nairobi');
```
Comment (kilemensi) [inline public/js/counties.js]: Shouldn't `results` be passed to `countyGraphChange`? Or how does it "know" where to get data?

File: public/js/countyGraph.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
+const graph = document.getElementById('graph').getContext('2d');
+Chart.defaults.global.defaultFontSize = 10;
+
+const countyGraphChange = async (county) => {
+  const kenyaCounties = [
+    'Nairobi',
```
Comment (kilemensi) [inline public/js/countyGraph.js]: Why are we hard-coding this list here and also fetch the same list via `await getCounties()` below?

File: public/js/countyGraph.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  });
+
+  const newArr = new Array(47 - Object.keys(countiesMap).length).fill(0); // populate the rest of the counties with 0 nodes
+
+  // Array of counties
+  const countiesArr = [
+    ...new Set([...Object.keys(countiesMap), ...kenyaCounties]),
+  ];
+  // Array of nodes in counties
+  const countiesDataArr = [...Object.values(countiesMap), ...newArr];
```
Comment (kilemensi) [inline public/js/countyGraph.js]: Not sure about this computation here; It looks to me like you're counting cities in a county and not nodes (that should come from `results`) in a city in a county.

File: public/js/countyGraph.js [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+  new Chart(graph, {
```
Comment (kilemensi) [inline public/js/countyGraph.js]: Where are we assigning this?

File: public/js/events.js (hunk 3 above)
Comment (kilemensi) [reply public/js/events.js]: I don't see how

File: src/pages/api/data.js [hunk 10]
```diff
@@ -1,8 +1,19 @@
 import { fetchAllNodes } from 'api';
 
 export default async function handler(req, res) {
+  const daysAgo = req.query.length;
```
Comment (kilemensi) [inline src/pages/api/data.js]: You need a default value in case `req.query.length` is undefined.

File: src/pages/api/data.js [hunk 11]
```diff
@@ -1,8 +1,19 @@
...
+  const date = new Date();
+  const dateRange = date.setDate(date.getDate() - daysAgo);
+  let ISOFormat;
+
+  if (dateRange) {
+    ISOFormat = new Date(dateRange).toISOString();
```
Comment (kilemensi) [inline src/pages/api/data.js]: ISOFormat?

**Review by kilemensi — COMMENTED** [review]
👍🏾 

---

Don't you have eslint or something setup up? 👇🏾 is sample of how this code looks in my VS Code
Screenshot from 2021-06-21 12-10-08

File: public/js/counties.js (hunk 6 above)
Comment (gertie-sheshe) [reply public/js/counties.js]: Order of JS files in the HTML file, result data is a variable that can be accessed in this scope @kilemensi

Comment (kilemensi) [reply public/js/counties.js]: I can see that @gertie-sheshe but when have we ever written function code that depends on global variables?

File: public/js/countyGraph.js (hunk 8 above)
Comment (gertie-sheshe) [reply public/js/countyGraph.js]: I calculated the nodes per county (city is what I use to get the county it belongs to) from results [here](https://github.com/CodeForAfrica/sensors.AFRICA-Dashboard-KE-HAP/pull/10/files#diff-6cbf18c3bcff269253eee101b966c4d44588e940e870272dc883b2b98bdcd0a6R59) and put them in the object `countiesMap` @kilemensi

File: public/js/countyGraph.js (hunk 7 above)
Comment (gertie-sheshe) [reply public/js/countyGraph.js]: Fixed

File: src/pages/api/data.js (hunk 11 above)
Comment (gertie-sheshe) [reply src/pages/api/data.js]: ...the [date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString) string returned is in ISO Format. I can rename it though @kilemensi

File: src/pages/api/data.js (hunk 10 above)
Comment (gertie-sheshe) [reply src/pages/api/data.js]: Fixed

File: public/js/counties.js (hunk 5 above)
Comment (gertie-sheshe) [reply public/js/counties.js]: Changed

File: public/js/countyGraph.js (hunk 9 above)
Comment (gertie-sheshe) [reply public/js/countyGraph.js]: Fixed

File: public/js/counties.js (hunk 6 above)
Comment (gertie-sheshe) [reply public/js/counties.js]: Used a different approach @kilemensi Take a look

File: public/js/events.js (hunk 3 above)
Comment (gertie-sheshe) [reply public/js/events.js]: Made a change @kilemensi take a look

**Review by kilemensi — APPROVED** [review]

**Pull request merged**